File: class_session/2_sept/BST_Creation.py


class Node_Tree:
    def __init__(self,data):
        self.data=data
        self.left=None
        self.right=None

def inorder(root):
    if root is not None:
        inorder(root.left)
        print(root.data,end=" ")
        inorder(root.right)

def insert(root,value):
    if root is None:
        return Node_Tree(value)
    if root.data>value:
        root.left=insert(root.left,value)
    if root.data<value:
        root.right=insert(root.right,value)
    return root  
def search(root,key_ele):
    if root is None or root.data==key_ele:
        return root
    if root.data>key_ele:
        return search(root.left,key_ele)
    else:
        return search(root.right,key_ele)
def isValid(root,min,max):
    if root is None:
        return True
    if root.data<min and root.data>max:
        return False
    return isValid(root.left,min,root.data) and isValid(root.right,root.data,max)

def min_ele(root):
    temp=root
    while temp.left is not None:
        temp=temp.left
    return temp.data
def max_ele(root):
    temp=root
    while temp.right is not None:
        temp=temp.right
    return temp.data

def predecessor(root):
    return max_ele(root.left)
def successor(root):
    return min_ele(root.right)

root=None
root=insert(root,12)
root=insert(root,8)
root=insert(root,4)
root=insert(root,25)
print("tree travelsing")

print("\ninorder=")
inorder(root)
print(search(root,56))

min=float('-inf')
max=float('inf')
print("is valid BST=",isValid(root,min,max))

# min ele of bst
print("Min ele of BST=",min_ele(root))

# max ele of bst
print("Max ele of BST=",max_ele(root))

# Predecessor of BST
print("Predecessor of BST=",predecessor(root))
print("Successor of BST=",successor(root))

File: class_session/2_sept/BST_creation.java

class Node_Tree
{
    public int data;
    public Node_Tree left;
    public Node_Tree right;
    public Node_Tree(int data)
    {
        this.data=data;
        left=null;
        right=null;
    }
}
public class BST_creation
{
    public static void main(String args[])
    {
        Node_Tree root=null;
        root=insert(root,12);
        root=insert(root,8);
        root=insert(root,25);
        root=insert(root,4);
        root=insert(root,18);
        root=insert(root,32);
        root=insert(root,10);
        //display the elements
        inorder(root);
        //search
        System.out.println("\nele is present:="+search(root,25));
        //valid BST or not
        int min=Integer.MIN_VALUE;
        int max=Integer.MAX_VALUE;
        System.out.println("is it valid bst="+isValid(root,min,max));

        //min value of BST
        System.out.println("min ele of BST="+min(root));
        //max value of BST
        System.out.println("max ele of BST="+max(root));

        //Predecessor and successor 
        System.out.println("Predecessor of BST="+predecessor(root));
        System.out.println("Successor of BST="+successor(root));
    }
    public static Node_Tree insert(Node_Tree root, int value)
    {
        if(root==null)
            return new Node_Tree(value);
        if(root.data<value)
            root.right=insert(root.right,value);
        if(root.data>value)
            root.left=insert(root.left,value);
        return root;

    }
    public static void inorder(Node_Tree root)
    {
        if(root!=null)
        {
            inorder(root.left);
            System.out.print(root.data+"->");
            inorder(root.right);
        }
    }

    public static Node_Tree search(Node_Tree root,int key_ele)
    {
        if(root==null || root.data==key_ele)
            return root;
        if(root.data>key_ele)
            return search(root.left,key_ele);
        else 
           return search(root.right,key_ele);
    }

    public static boolean isValid(Node_Tree root,int min, int max)
    {
        if(root==null)
            return true;
        if(root.data<min || root.data>max)
            return false;
        return isValid(root.left,min,root.data)&&isValid(root.right,root.data,max);
    }

    public static int min(Node_Tree root)
    {
        Node_Tree temp=root;
        while(temp.left!=null)
        {
            temp=temp.left;
        }
        return temp.data;
    }
     public static int max(Node_Tree root)
    {
        Node_Tree temp=root;
        while(temp.right!=null)
        {
            temp=temp.right;
        }
        return temp.data;
    }

    public static int predecessor(Node_Tree root)
    {
        return max(root.left);
    }

    public static int successor(Node_Tree root)
    {
        return min(root.right);
    }
}
